[X86] Add MI-layer routine for getting the index of the first address operand, NFC

nmosier opened the pull request:
Add the MI-layer routine X86::getFirstAddrOperandIdx(), which returns the index of the first address operand of a MachineInstr (or -1 if there is none).

X86II::getMemoryOperandNo(), the existing MC-layer routine used to obtain the index of the first address operand in a 5-operand X86 memory reference, is incomplete: it does not handle pseudo-instructions like TCRETURNmi, resulting in security holes in the mitigation passes that use it (e.g., x86-slh and x86-lvi-load).

X86::getFirstAddrOperandIdx() handles both pseudo and real instructions and is thus more suitable for most use cases than X86II::getMemoryOperandNo(), especially in mitigation passes like x86-slh and x86-lvi-load. For this reason, this patch replaces all uses of X86II::getMemoryOperandNo() with X86::getFirstAddrOperandIdx() in the aforementioned mitigation passes.

nmosier committed:
[X86] Add MI-layer routine for getting the index of the first address operand

Add the MI-layer routine X86::getFirstAddrOperandIdx(), which returns the index
of the first address operand of a MachineInstr (or -1 if there is none).

X86II::getMemoryOperandNo(), the existing MC-layer routine used to obtain the index of the first address
operand in a 5-operand X86 memory reference, is incomplete:
it does not handle pseudo-instructions like TCRETURNmi, resulting
in missed optimizations (e.g., x86-optimize-LEAs)
and security holes (e.g., x86-slh and x86-lvi-load)
for the passes that use it.

X86::getFirstAddrOperandIdx() handles both pseudo and real instructions
and is thus more suitable for most use cases than X86II::getMemoryOperandNo().
For this reason, this patch replaces most uses of the latter with the former.

phoebewang commented on llvm/lib/Target/X86/X86SpeculativeLoadHardening.cpp:
```diff
@@ -1365,7 +1358,7 @@ void X86SpeculativeLoadHardeningPass::tracePredStateThroughBlocksAndHarden(
         if (EnablePostLoadHardening && X86InstrInfo::isDataInvariantLoad(MI) &&
             !isEFLAGSDefLive(MI) && MI.getDesc().getNumDefs() == 1 &&
-            MI.getOperand(0).isReg() &&
+            MI.getOperand(0).isReg() && MI.getOperand(0).getReg().isVirtual() &&
```
Do not put unrelated change in one patch.

nmosier replied:
Removed. That got in there by mistake (it's from a different patch).

phoebewang commented on llvm/test/CodeGen/X86/vaargs.ll:
```diff
@@ -1,4 +1,5 @@
-; RUN: llc -verify-machineinstrs -mcpu=corei7-avx %s -o - | FileCheck %s --check-prefix=CHECK --check-prefix=NO-FLAGS
+; RUN: llc -verify-machineinstrs -mcpu=corei7-avx -disable-x86-lea-opt %s -o - | FileCheck %s --check-prefix=CHECK --check-prefix=NO-FLAGS
+; Disable the x86-optimize-LEAs pass, since it may introduce an intermediate LEA instruction that changes the base register of the vmovaps instructions from %rsp to something else.
```
Why do we need the change here? I understand the patch is NFC.

nmosier replied:
I reverted the changes in X86OptimizeLEAs.cpp (it uses X86II::getMemoryOperandNo again) and the changes to this test. It looks like X86OptimizeLEAs can't handle pseudo instructions like VASTART_SAVE_XMM_REGS that may expand to multiple memory accesses at different displacements.

phoebewang commented on llvm/lib/Target/X86/X86DiscriminateMemOps.cpp:
```diff
@@ -129,7 +129,7 @@ bool X86DiscriminateMemOps::runOnMachineFunction(MachineFunction &MF) {
   for (auto &MBB : MF) {
     for (auto &MI : MBB) {
-      if (X86II::getMemoryOperandNo(MI.getDesc().TSFlags) < 0)
+      if (X86::getFirstAddrOperandIdx(MI) < 0)
```
Maybe still use `getMemoryOperandNo`?

nmosier replied:
Reverted.

nmosier committed:
Address comments

Revert X86DiscriminateMemOps, X86OptimizeLEAs.
Remove extraneous change in X86SpeculativeLoadHardening.
Revert vaargs.ll test, since X86OptimizeLEAs was reverted.

phoebewang commented on llvm/lib/Target/X86/X86AvoidStoreForwardingBlocks.cpp:
```diff
@@ -289,10 +289,8 @@ static unsigned getYMMtoXMMStoreOpcode(unsigned StoreOpcode) {
 static int getAddrOffset(const MachineInstr *MI) {
-  const MCInstrDesc &Descl = MI->getDesc();
-  int AddrOffset = X86II::getMemoryOperandNo(Descl.TSFlags);
+  const int AddrOffset = X86::getFirstAddrOperandIdx(*MI);
```
The `const` doesn't work here. Change the function to `static const int` or remove it.

nmosier replied:
Is it just a stylstic preference? Or why doesn't declaring `AddrOffset` as `const` work here?

phoebewang replied:
We usually don't use `const` for integer tpye. I think you may want to highlight `AddrOffset` is the final offset with bias and don't change it anymore by using `const`. In that way, it doesn't work here, because the function doesn't consume `AddrOffset` but return it in a non-const int.

nmosier replied:
Thanks, fixed.

phoebewang reviewed (approved):
LGTM except one nit.

nmosier committed:
const int -> int

KanRobert commented:
> it does not handle pseudo-instructions like TCRETURNmi, resulting in missed optimizations (e.g., x86-optimize-LEAs)

Why there is no test update for this?

phoebewang commented:
> > it does not handle pseudo-instructions like TCRETURNmi, resulting in missed optimizations (e.g., x86-optimize-LEAs)
> 
> Why there is no test update for this?

The change was reverted, but the description doesn't update.

KanRobert commented on llvm/lib/Target/X86/X86InstrInfo.cpp:
```diff
@@ -3463,6 +3463,44 @@ bool X86::isX87Instruction(MachineInstr &MI) {
   return false;
 }
 
+int X86::getFirstAddrOperandIdx(const MachineInstr &MI) {
+  const auto isMemOp = [](const MCOperandInfo &OpInfo) -> bool {
+    return OpInfo.OperandType == MCOI::OPERAND_MEMORY;
+  };
+
+  const MCInstrDesc &Desc = MI.getDesc();
+
+  // An instruction cannot have a memory reference if it has fewer than
+  // AddrNumOperands (= 5) explicit operands.
+  if (Desc.getNumOperands() < X86::AddrNumOperands) {
+    assert(none_of(Desc.operands(), isMemOp) &&
+           "Expected no operands to have OPERAND_MEMORY type!");
+    return -1;
+  }
+
+  // The first operand with type OPERAND_MEMORY indicates the start of a memory
+  // reference. We expect the following AddrNumOperand-1 operands to also have
+  // OPERAND_MEMORY type.
+  for (unsigned i = 0; i <= Desc.getNumOperands() - X86::AddrNumOperands; ++i) {
+    if (Desc.operands()[i].OperandType == MCOI::OPERAND_MEMORY) {
+      assert(std::all_of(Desc.operands().begin() + i,
+                         Desc.operands().begin() + i + X86::AddrNumOperands,
+                         isMemOp) &&
+             "Expected all five operands in the memory reference to have "
+             "OPERAND_MEMORY type!");
+      return i;
+    }
+  }
+
+  // Fall back to the MC-layer routine, which only handles real (not pseudo)
+  // insturctions.
+  const int FallbackMemOpNo = X86II::getMemoryOperandNo(Desc.TSFlags);
+  if (FallbackMemOpNo >= 0)
+    return FallbackMemOpNo + X86II::getOperandBias(Desc);
+
+  return -1;
+}
```
This implementation has some flaws

1. It increases compile-time for non-pseduo instructions
2. It delays the time when the bug (wrong format) is exposed

Couldn't we can check pseduo and fall back to MC-layer first?

KanRobert commented:
> > > it does not handle pseudo-instructions like TCRETURNmi, resulting in missed optimizations (e.g., x86-optimize-LEAs)
> > 
> > 
> > Why there is no test update for this?
> 
> The change was reverted, but the description doesn't update.

So it does allow more optimization to happen?

nmosier committed:
Fall back to MC-layer X86II::getMemoryOperandNo() first

nmosier committed:
Revert uses of X86II::getMemoryOperandNo() in optimization passes

KanRobert commented on llvm/lib/Target/X86/X86InstrInfo.cpp:
```diff
@@ -3463,6 +3463,50 @@ bool X86::isX87Instruction(MachineInstr &MI) {
+    int MemRefIdx = X86II::getMemoryOperandNo(Desc.TSFlags);
+    if (MemRefIdx >= 0)
+      return MemRefIdx + X86II::getOperandBias(Desc);
+    assert(none_of(Desc.operands(), isMemOp) &&
+           "Got false negative from X86II::getMemoryOperandNo()!");
```
Wrap these in 

#ifdef EXPENSIVE_CHECKS
#endif

nmosier replied:
Done

KanRobert commented on llvm/lib/Target/X86/X86InstrInfo.cpp:
```diff
@@ -3463,6 +3463,50 @@ bool X86::isX87Instruction(MachineInstr &MI) {
+  // operands (slow case). An instruction cannot have a memory reference if it
+  // has fewer than AddrNumOperands (= 5) explicit operands.
+  if (Desc.getNumOperands() < X86::AddrNumOperands) {
+    assert(none_of(Desc.operands(), isMemOp) &&
+           "Expected no operands to have OPERAND_MEMORY type!");
```
Wrap these in 

#ifdef EXPENSIVE_CHECKS
#endif

nmosier replied:
Done

KanRobert commented on llvm/lib/Target/X86/X86InstrInfo.cpp:
```diff
@@ -3463,6 +3463,50 @@ bool X86::isX87Instruction(MachineInstr &MI) {
+  // OPERAND_MEMORY type.
+  for (unsigned i = 0; i <= Desc.getNumOperands() - X86::AddrNumOperands; ++i) {
+    if (Desc.operands()[i].OperandType == MCOI::OPERAND_MEMORY) {
+      assert(std::all_of(Desc.operands().begin() + i,
+                         Desc.operands().begin() + i + X86::AddrNumOperands,
+                         isMemOp) &&
+             "Expected all five operands in the memory reference to have "
+             "OPERAND_MEMORY type!");
```
Wrap these in 

#ifdef EXPENSIVE_CHECKS
#endif

nmosier replied:
Done

nmosier committed:
Expensive checks

KanRobert reviewed (approved):
LGTM

nmosier commented:
Thanks! Can you merge it (as I don't have commit access)?

KanRobert commented:
> Thanks! Can you merge it (as I don't have commit access)?

Sure
